pkg/repl/client.go

```diff
@@ -562,23 +564,23 @@ func (c *Client) BlockWait(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	ctx, cancel := context.WithTimeout(ctx, DefaultTimeout)
-	defer cancel()
-	for ctx.Err() == nil { // end when the timeout is reached.
-		// Inject some noise into the binlog stream
-		// This is to ensure that we are past the current position
-		// without an off-by-one error
-		if err := dbconn.Exec(ctx, c.db, "FLUSH BINARY LOGs"); err != nil {
-			break // error flushing binary logs
-		}
-		// fetch the bufferPos under a mutex.
-		if c.getBufferedPos().Compare(targetPos) >= 0 {
-			return nil // we are up to date!
+	c.logger.Infof("waiting to catch up to source position: %v, current position is: %v", targetPos, c.getBufferedPos())
+	timer := time.NewTimer(DefaultTimeout)
+	for {
+		select {
+		case <-timer.C:
+			return fmt.Errorf("timed out waiting to catch up to source position: %v, current position is: %v", targetPos, c.getBufferedPos())
+		default:
+			if err := dbconn.Exec(ctx, c.db, "FLUSH BINARY LOGS"); err != nil {
+				break // error flushing binary logs
+			}
+			if c.getBufferedPos().Compare(targetPos) >= 0 {
+				return nil // we are up to date!
+			}
+			// We are not caught up yet, so we need to wait.
+			time.Sleep(100 * time.Millisecond)
```

Comment:
This refactors the code to be the same as canal, which we used previously.

I don't think this was the cause of the timeouts, but when I was trying to debug this - this struck me as sensitive code, and the canal implementation is well tested.